pkg/asset/installconfig/gcp/services.go

```diff
@@ -4,81 +4,97 @@ import (
+// for WithEndpoint or BasePath override endpoint options.
+func FormatGCPEndpointList(endpoints []configv1.GCPServiceEndpoint, input FormatGCPEndpointInput) ([]configv1.GCPServiceEndpoint, error) {
 	// The endpoints are modified to include the path
 	modifiedEndpoints := []configv1.GCPServiceEndpoint{}
 	for _, se := range endpoints {
-		var err error
-		var formattedURL string
-		switch se.Name {
-		case configv1.GCPServiceEndpointNameCloudResource:
-			formattedURL, err = FormatGCPEndpoint(se.URL, CloudResourceManagerNameGCP, ServiceVersionGCP3)
-		case configv1.GCPServiceEndpointNameCompute:
-			formattedURL, err = FormatGCPEndpoint(se.URL, ComputeServiceNameGCP, ServiceVersionGCP1)
-		case configv1.GCPServiceEndpointNameDNS:
-			formattedURL, err = FormatGCPEndpoint(se.URL, DNSServiceNameGCP, ServiceVersionGCP1)
-		//case configv1.GCPServiceEndpointNameFile:
-		//	formattedURL, err = FormatGCPEndpoint(se.URL, FileServiceNameGCP, ServiceVersionGCP1)
-		case configv1.GCPServiceEndpointNameIAM:
-			formattedURL, err = FormatGCPEndpoint(se.URL, IAMServiceNameGCP, ServiceVersionGCP1)
-		//case configv1.GCPServiceEndpointNameServiceUsage:
-		//	formattedURL, err = FormatGCPEndpoint(se.URL, ServiceUsageNameGCP, ServiceVersionGCPBeta)
-		case configv1.GCPServiceEndpointNameStorage:
-			formattedURL, err = FormatGCPEndpoint(se.URL, StorageServiceNameGCP, ServiceVersionGCP1)
+		formattedURL, err := FormatGCPEndpoint(se.Name, se.URL, input)
+		if err != nil {
+			return nil, fmt.Errorf("failed to format GCP service endpoint %s: %w", se.Name, err)
 		}
 
-		if err != nil || formattedURL == "" {
-			return nil, fmt.Errorf("failed to format GCP Service Endpoint URL %s: %w", se.URL, err)
-		}
+		logrus.Debugf("Formatted GCP service endpoint %s: %s", se.Name, formattedURL)
 		modifiedEndpoints = append(modifiedEndpoints, configv1.GCPServiceEndpoint{Name: se.Name, URL: formattedURL})
 	}
 	return modifiedEndpoints, nil
 }
 
 // FormatGCPEndpoint will format the endpoint to ensure that the string is in the format that would be
 // accepted by both options (WithEndpoint and BasePath override).
-func FormatGCPEndpoint(endpoint string, service ServiceNameGCP, version ServiceVersionGCP) (string, error) {
+func FormatGCPEndpoint(service configv1.GCPServiceEndpointName, endpoint string, input FormatGCPEndpointInput) (string, error) {
 	endpointURL, err := url.Parse(endpoint)
 	if err != nil {
 		return "", fmt.Errorf("failed to parse GCP Service Endpoint URL %s: %w", endpoint, err)
 	}
 
+	if endpointURL.Host == "" {
+		logrus.Debugf("GCP endpoint did not set a host, setting host to %s", endpoint)
+		endpointURL.Host = endpoint
+		endpointURL.Path = ""
+	}
+
+	endpointURL.Path = strings.TrimSuffix(endpointURL.Path, "/")
 	endpointURL.Scheme = "https"
-	endpointURL.Path = fmt.Sprintf("/%s/%s/", service, version)
+	if !input.SkipPath {
+		switch service {
+		case configv1.GCPServiceEndpointNameCloudResource:
```

Comment:
Nit: can there be a comma separated list of services that all use `endpointURL.Path = "/"` ? Makes code more readable.

Reply:
@sadasu The skip path is expected behavior because there is at least one cluster component that requires that the path not be set. As far as the comma separated list I am not sure I understand. The install config provides information about the services that can be overridden. Why would I want to allow multiple entries for a single service that seems confusing.

Reply:
> As far as the comma separated list I am not sure I understand. 

```
switch service {
case configv1.GCPServiceEndpointNameCloudResource, configv1.GCPServiceEndpointNameDNS, configv1.GCPServiceEndpointNameFile, configv1.GCPServiceEndpointNameIAM, configv1.GCPServiceEndpointNameServiceUsage:
        endpointURL.Path = "/"
case xxxxx:
default:
}
```